File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> brackets = new HashMap<>(); 
        brackets.put('(', ')'); 
        brackets.put('{', '}'); 
        brackets.put('[', ']');
        // stack
        Stack<Character> stack = new Stack<>(); 
        for(char c : s.toCharArray()) {
            if(!brackets.containsKey(c)) {
                if(!stack.isEmpty()) {
                    if(brackets.get(stack.peek()) != c) {
                        return false; 
                    } else
                        stack.pop(); 
                } else
                    return false; 
            } else 
                stack.push(c); 
        }
        return stack.isEmpty(); 
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for char in s:
            if char in "([{":
                stack.append(char)
            else: 
                if len(stack) == 0:
                    return False
                elif char == ")" and stack[-1] != "(" or char == "]" and stack[-1] != "[" or char == "}" and stack[-1] != "{":
                    return False
                else: 
                    stack.pop()
        return len(stack) == 0
        